editor/src/marionette/preview.ts

```diff
@@ -0,0 +1,357 @@
+    }
+
+    public addVariable(id: string, type: VariableType, value: Value) {
+        this._varInstances[id] = new VarInstance(type, value);
+    }
+
+    public removeVariable(id: string) {
+        delete this._varInstances[id];
+    }
+
+    public updateVariable(id: string, value: Value) {
+        this._varInstances[id].value = value;
+    }
+
+    protected createMotionEval(motion: Motion) {
+        const {
+            _root: root,
+            _blendBuffer: blendBuffer,
+        } = this;
+
+        const bindContext: BindContext = {
+            getVar: this._getVar.bind(this),
+        };
+
+        const motionEval = motion[createEval]({
+            node: root,
+            blendBuffer,
+            getVar: (...args) => {
+                return bindContext.getVar(...args);
+            },
+        });
+
+        return motionEval;
+    }
+
+    private _root: Node;
+
+    private _blendBuffer: BlendStateBuffer = new LegacyBlendStateBuffer();
+
+    private _varInstances: Record<string, VarInstance> = {};
+
+    private _getVar(id: string): VarInstance | undefined {
+        return this._varInstances[id];
+    }
+}
+
+export interface MotionPreviewerTimelineStats {
+    timeLineLength: number;
+}
+
+export class MotionPreviewer extends AnimationGraphPartialPreviewer {
+    get timelineStats(): Readonly<MotionPreviewerTimelineStats> {
+        if (this._timelineStatsDirty) {
+            this._updateTimelineStats();
+            this._timelineStatsDirty = false;
+        }
+        return this._timelineStats;
+    }
+
+    /**
+     * Gets an iterable to the weights of each motion(that has runtime ID).
+     */
+    public queryWeights(): Iterable<[RuntimeID, number]> {
+        const getWeightsRecursive = function* (motionEval: MotionEval, weight: number): Iterable<[RuntimeID, number]> {
+            if (typeof motionEval.runtimeId !== 'undefined') {
+                yield [motionEval.runtimeId, weight];
+            }
+
+            if (motionEval instanceof AnimationBlendEval) {
+                const nChild = motionEval.childCount;
+                for (let iChild = 0; iChild < nChild; ++iChild) {
+                    const childMotionEval = motionEval.getChildMotionEval(iChild);
+                    const childWeight = motionEval.getChildWeight(iChild);
+                    if (childMotionEval) {
+                        for (const child of getWeightsRecursive(childMotionEval, childWeight)) {
+                            yield child;
+                        }
+                    }
+                }
+            }
+
+            return;
+        };
+
+        if (this._motionEval) {
+            return getWeightsRecursive(this._motionEval, 1.0);
+        }
+
+        return [];
+    }
+
+    public setMotion(motion: Motion) {
+        this._motionEval = super.createMotionEval(motion);
+        this._timelineStatsDirty = true;
+    }
+
+    public setTime(time: number) {
+        this._time = time;
```

Comment:
should set timeline stats dirty?

Reply:
Time does not belong to timeline stats as suggested in https://github.com/shrinktofit/engine/blob/AnimationGraphEditorAPI/editor/src/marionette/preview.ts#L161